### PR TITLE
Shader load time performance optimiziations

### PR DESCRIPTION
**Pull request by jherico**
* Cache shader binaries
* Optimize loading shader uniforms
* Add unit test for benchmarking shader load performance
* Fixup unit tests to use test-utils library

## Testing

This PR will cache the shaders we use as binary data after the first run.  The cache will be located in the application data directory, in the case of this PR, on Windows that would be `~/AppData/High Fidelity - PR13082/Interface/shaders/cache.json`.  

### Windows testing

Testing will require both an nVidia and an AMD based machine

* On the nVidia machine, install the PR build.  Verify that the above file does not exist.
* Run Interface
  * Verify the file still does not exist
  * Verify rendering is normal
  * Switch between desktop and HMD or stereo display modes.  Verify rendering is normal
* Exit Interface
* Verify the shader cache file exists and is populated (it should be about 10MB on a windows machine with recent nVidia drivers)
* Restart interface.  
  * The application should move from the grey screen to a rendered scene faster than before (although the difference might be very slight)
  * Re-follow the same steps as before for validating rendering
* Exit interface
* Copy the shader cache from the above location to a flash drive.
* On the AMD machine, install the PR
* Verify the shader cache does not exist
* Copy the shader cache from the flash drive to the AMD machine
* Run interface (on the AMD system)
  * Verify that interface starts normally.  
  * Run the same rendering validation steps for desktop and HMD/Stereo
* Exit interface
* Restart interface (on the AMD system)
  * The application should move from the grey screen to a rendered scene faster than before (although the difference might be very slight)
  * Run the same rendering validation steps for desktop and HMD/Stereo
* Exit interface

### Mac testing

Install and test run the PR build.  The Mac does not support cached binary shaders for OpenGL, so the file may or may not be created, but will likely be an empty JSON object if it exits.

* Verify that the application starts up and shuts down successfully several times
* The application startup time should be significantly reduced compared to the current master build.  

### Unit test tool

https://hifi-public.s3.amazonaws.com/austin/ShaderLoadTest.zip

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9453/app_13082_PR_d87b35583806b0d53021c1036faa95abf0f1a811.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9457/app_13082_PR_4c96454258247bc158eade48f4f07e3d5b8e3214.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9459/app_13082_PR_920fc9097618524fc162f9eca6c5f9e40121d781.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9459/HighFidelity-Beta-PR13082-920fc9097618524fc162f9eca6c5f9e40121d781.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9459/HighFidelity-Beta-Interface-PR13082-920fc9097618524fc162f9eca6c5f9e40121d781.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9459/HighFidelity-Beta-PR13082-920fc9097618524fc162f9eca6c5f9e40121d781.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9459/HighFidelity-Beta-Interface-PR13082-920fc9097618524fc162f9eca6c5f9e40121d781.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/9464/app_13082_PR_c705840f77dfc13bb900dbf4113fca2212f257c8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9464/HighFidelity-Beta-PR13082-c705840f77dfc13bb900dbf4113fca2212f257c8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/9464/HighFidelity-Beta-Interface-PR13082-c705840f77dfc13bb900dbf4113fca2212f257c8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9464/HighFidelity-Beta-PR13082-c705840f77dfc13bb900dbf4113fca2212f257c8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/9464/HighFidelity-Beta-Interface-PR13082-c705840f77dfc13bb900dbf4113fca2212f257c8.exe

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -2492,6 +2492,7 @@ void Application::cleanupBeforeQuit() {
     }
 
     _window->saveGeometry();
+    _gpuContext->shutdown();
```
Comment (jherico) [inline interface/src/Application.cpp]: Ensure we persist the shader binary cache even if we don't exit the application normally and the `_gpuContext` isn't destroyed.

File: libraries/gl/src/gl/GLShaders.cpp [hunk 2]
```diff
@@ -4,13 +4,44 @@
 
 namespace gl {
 
+    void Uniform::load(GLuint glprogram, int index) {
```
Comment (jherico) [inline libraries/gl/src/gl/GLShaders.cpp]: The current 4.1 backend does this parsing of all uniforms in a shader more than once.  This PR modifies the GL shader object to contain storage for all the uniforms.  Instead of iterating over all the uniforms using GL calls multiple times, we now load all the uniforms and store them in the shader object up front, then the code that previously used `glGetActiveUniform`  now has an existing vector to iterate instead.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10031/HighFidelity-Beta-PR13082-e30761e6592a36405f55e293803bef3c2befaccf.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10032/HighFidelity-Beta-PR13082-968b1fc2198c7742771c1e11c24acc378be97cd4.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10039/HighFidelity-Beta-PR13082-f7cc38fa2d99bc2da538915e121629f864a65f37.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10039/HighFidelity-Beta-PR13082-f7cc38fa2d99bc2da538915e121629f864a65f37.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10039/HighFidelity-Beta-Interface-PR13082-f7cc38fa2d99bc2da538915e121629f864a65f37.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10039/HighFidelity-Beta-PR13082-f7cc38fa2d99bc2da538915e121629f864a65f37.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10039/HighFidelity-Beta-Interface-PR13082-f7cc38fa2d99bc2da538915e121629f864a65f37.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10063/HighFidelity-Beta-PR13082-bdd71b30855ab0b33a31a8722d59a6874df9871a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10063/HighFidelity-Beta-PR13082-bdd71b30855ab0b33a31a8722d59a6874df9871a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10063/HighFidelity-Beta-Interface-PR13082-bdd71b30855ab0b33a31a8722d59a6874df9871a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10063/HighFidelity-Beta-PR13082-bdd71b30855ab0b33a31a8722d59a6874df9871a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10063/HighFidelity-Beta-Interface-PR13082-bdd71b30855ab0b33a31a8722d59a6874df9871a.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10089/HighFidelity-Beta-PR13082-7e163258050b260aab6c26fd096d456200584131.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10089/HighFidelity-Beta-PR13082-7e163258050b260aab6c26fd096d456200584131.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10089/HighFidelity-Beta-Interface-PR13082-7e163258050b260aab6c26fd096d456200584131.dmg

File: libraries/networking/src/ResourceCache.cpp [hunk 3]
```diff
@@ -218,8 +218,8 @@ ScriptableResource* ResourceCache::prefetch(const QUrl& url, void* extra) {
 }
 
 ResourceCache::ResourceCache(QObject* parent) : QObject(parent) {
-    auto nodeList = DependencyManager::get<NodeList>();
-    if (nodeList) {
+    if (DependencyManager::isSet<NodeList>()) {
+        auto nodeList = DependencyManager::get<NodeList>();
```
Comment (samcake) [inline libraries/networking/src/ResourceCache.cpp]: WHy is that needed in this pr ?

**Comment by samcake** [comment]
@jherico  COuld you provide a build version of the unit test for Michael to be able to run these too, thanks

File: libraries/networking/src/ResourceCache.cpp (hunk 3 above)
Comment (jherico) [reply libraries/networking/src/ResourceCache.cpp]: It triggers logspam in my unit tests, because if you call `get` without checking `isSet` first, the dependency manager emits a warning.

**Comment by Qliemillar** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10164/HighFidelity-Beta-PR13082-7e163258050b260aab6c26fd096d456200584131.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10164/HighFidelity-Beta-Interface-PR13082-7e163258050b260aab6c26fd096d456200584131.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10170/HighFidelity-Beta-PR13082-7e163258050b260aab6c26fd096d456200584131.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10170/HighFidelity-Beta-PR13082-7e163258050b260aab6c26fd096d456200584131.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10170/HighFidelity-Beta-Interface-PR13082-7e163258050b260aab6c26fd096d456200584131.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Stopped Testing

**Comment by qligmaloney** [comment]
Currently testing

**Comment by qligmaloney** [comment]
PASS @conklin94122 Interface renders faster, shader cache is generated. AMD system renders as expected with Nvidia cache. Shader cache JSON generated on Mac but is empty
